### PR TITLE
refactor: use dataclasses instead of NamedTuples

### PR DESCRIPTION
**Pull request by henryiii**
Yesterday's MyPy 0.950 fixed a bug with dataclasses holding Callables (the one in the bashlex file), so this is now supported (via static typing, it worked before but failed a typecheck). NamedTuples should not be used for new API, they are intended to be a backwards compatibility thing (they add all of tuple syntax too). 

See https://twitter.com/brettsky/status/1516493479282503687?s=20&t=2i7N_WHyCCDZ7GeiCqmgOA and other places.

### CODE REVIEW

File: cibuildwheel/__main__.py [hunk 1]
```diff
@@ -107,7 +107,7 @@ def main() -> None:
         help="Enable pre-release Python versions if available.",
     )
 
-    args = parser.parse_args(namespace=CommandLineArguments())
```
Comment (joerick) [inline cibuildwheel/__main__.py:110]: I guess this'll give us a runtime failure if the args don't match the object, which would help us catch mismatches.

**Review by joerick — APPROVED** [review]
Looks good!

**Comment by joerick** [comment]
Hm. CI failed for some reason. Oh well, it's not critical to get this in before the release.

**Comment by henryiii** [comment]
That's okay, no rush. I can't actually see why it failed ATM.

**Comment by henryiii** [comment]
Seems to have been an CI infrastructure error. Rebuilding the failed job worked.

**Pull request merged**